include/contraction/contractionGraph.hpp

```diff
@@ -40,220 +40,304 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+        std::sort(
+            o_eids.begin(),
+            o_eids.end(),
+            [&](E lhs, E rhs) {
+                return -1*(this->graph)[lhs].id < -1*(this->graph)[rhs].id;
+            });
+        return o_eids;
+    }
+
+    /*!
+        @brief vertices with at least one contracted vertex
+        @result The vids Identifiers with at least one contracted vertex
+    */
+    Identifiers<int64_t> get_modified_vertices() {
+        Identifiers<int64_t> vids;
+        for (auto v : boost::make_iterator_range(
+            boost::vertices(this->graph))) {
+            if ((this->graph[v]).has_contracted_vertices()) {
+                vids += (this->graph[v]).get_id();
+            }
+        }
+        return vids;
+    }
+
+    /*!
+        @brief get the edge with minimum cost between two vertices
+        @param [in] u vertex_descriptor of source vertex
+        @param [in] v vertex_descriptor of target vertex
+        @return E: The edge descriptor of the edge with minimum cost
+    */
+    std::tuple<double, Identifiers<int64_t>, bool> get_min_cost_edge(V u, V v) {
+        E min_edge;
+        Identifiers<int64_t> contracted_vertices;
+        double min_cost = (std::numeric_limits<double>::max)();
+        bool found = false;
+
+        if (this->is_directed()) {
+            BGL_FORALL_OUTEDGES_T(u, e, this->graph, G) {
+                if (target(e, this->graph) == v) {
+                    contracted_vertices +=
+                        (this->graph[e]).get_contracted_vertices();
+                    if ((this->graph[e]).cost < min_cost) {
+                        min_cost = (this->graph[e]).cost;
+                        min_edge = e;
+                        found = true;
+                    }
+                }
+            }
+            return std::make_tuple(min_cost, contracted_vertices, found);
+        }
+
+        pgassert(this->is_undirected());
+        BGL_FORALL_OUTEDGES_T(u, e, this->graph, G) {
+            if (this->adjacent(u, e) == v) {
+                contracted_vertices +=
+                    (this->graph[e]).get_contracted_vertices();
+                if ((this->graph[e]).cost < min_cost) {
+                    min_cost = (this->graph[e]).cost;
+                    min_edge = e;
+                    found = true;
+                }
+            }
+        }
+        return std::make_tuple(min_cost, contracted_vertices, found);
+    }
+
+    /*!
+        @brief print the graph with contracted vertices of all vertices and edges
+    */
+    friend
+    std::ostream& operator << (std::ostream &os, const contractionGraph &g) {
+        EO_i out, out_end;
+        for (auto vi = vertices(g.graph).first;
+                vi != vertices(g.graph).second;
+                ++vi) {
+            if ((*vi) >= g.num_vertices()) break;
+            os << g.graph[*vi].id << "(" << (*vi) << ")"
+                << g.graph[*vi].get_contracted_vertices() << std::endl;
```

Comment:
```suggestion
                << g.graph[*vi].get_contracted_vertices() << "\n";
```

std::endl is only useful when you want to flush, so in almost all cases you want only "\n"